GCSProxyServer.go

```diff
@@ -0,0 +1,256 @@
+				return
+			}
+		default:
+			fmt.Printf("Unknown request type: %d\n", header.RequestType)
+			return
+		}
+	}
+
+	uploadSessionsMutex.Lock()
+	session.activeConnections--
+	if session.activeConnections == 0 && (session.isAborted || session.isCompleted) {
+		session.cancelFunc()
+		session.gcsClient.CancelUpload(ctx, session.uploadUrl)
+		delete(*uploadSessions, sessionKey)
+	}
+	uploadSessionsMutex.Unlock()
+}
+
+func createNewSession(ctx context.Context, uploadSessionsMutex *sync.Mutex, bucketName string, objectName string, session *UploadSession, uploadSessions *map[string]*UploadSession, sessionKey string) (*UploadSession, bool) {
+	sessionCtx, cancelFunc := context.WithTimeout(context.Background(), time.Hour)
+
+	gcsClient, err := utils.NewGcsClient(sessionCtx)
+	if err != nil {
+		uploadSessionsMutex.Unlock()
+		fmt.Printf("Failed to create GCS client: %v\n", err)
+		cancelFunc()
+		return nil, true
+	}
+	uploadUrl, err := gcsClient.NewUploadSession(sessionCtx, bucketName, objectName)
+	if err != nil {
+		uploadSessionsMutex.Unlock()
+		fmt.Printf("Failed to create upload session: %v\n", err)
+		cancelFunc()
+		return nil, true
+	}
+	session = &UploadSession{
+		uploadUrl:         uploadUrl,
+		gcsClient:         gcsClient,
+		resumeOffset:      0,
+		isAborted:         false,
+		isCompleted:       false,
+		activeConnections: 1,
+		bucketName:        bucketName,
+		objectName:        objectName,
+		sessionCtx:        sessionCtx,
+		cancelFunc:        cancelFunc,
+	}
+	(*uploadSessions)[sessionKey] = session
+
+	go func() {
+		<-sessionCtx.Done()
+		uploadSessionsMutex.Lock()
+		defer uploadSessionsMutex.Unlock()
+
+		if _, ok := (*uploadSessions)[sessionKey]; !ok {
+			return
+		}
+
+		session.isAborted = true
+		fmt.Printf("Session for %s/%s has timed out\n", bucketName, objectName)
+
+		if err := session.gcsClient.CancelUpload(context.Background(), session.uploadUrl); err != nil {
+			fmt.Printf("Error cancelling upload session: %v\n", err)
+		}
+
+		_ = session.gcsClient.CancelUpload(ctx, session.uploadUrl)
+		delete(*uploadSessions, sessionKey)
+	}()
+	return session, false
+}
+
+func handleWriteAt(ctx context.Context, conn net.Conn, session *UploadSession) error {
+	var writeAtReq writers.WriteAtRequestHeader
+	if err := binary.Read(conn, binary.BigEndian, &writeAtReq); err != nil {
+		return fmt.Errorf("failed to read WriteAtRequestHeader: %w", err)
+	}
+
+	dataSize := writeAtReq.ChunkEnd - writeAtReq.ChunkBegin
+	if dataSize <= 0 {
+		return errors.New("invalid data size")
+	}
+
+	if session.isAborted {
+		return errors.New("upload session is aborted")
+	}
+	if writeAtReq.ChunkBegin != session.resumeOffset {
+		return fmt.Errorf("chunk begin %d does not match resume offset %d", writeAtReq.ChunkBegin, session.resumeOffset)
+	}
+
+	limitedReader := io.LimitReader(conn, dataSize)
+	err := session.gcsClient.UploadObjectPart(ctx, session.uploadUrl, writeAtReq.ChunkBegin, limitedReader, dataSize, writeAtReq.IsLast != 0)
```

Comment:
This is good. This way you relay bytes from a client to GCS as soon as they arrive.